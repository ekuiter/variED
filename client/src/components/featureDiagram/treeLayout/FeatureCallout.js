import React from 'react';
import {Callout, DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {getSetting} from '../../../store/settings';
import contextualMenuItems from '../../contextualMenuItems';
import {CommandBar} from 'office-ui-fabric-react/lib/CommandBar';

export default props => {
    const {onDismiss, featureModel, featureName} = props,
        {gapSpace, width} = getSetting(props.settings, 'featureDiagram.treeLayout.overlay');
    const feature = featureModel && featureName ? featureModel.getFeature(featureName) : null;
    if (!feature)
        return null;
    return (
        <Callout target={props.nodeRef}
                 onDismiss={onDismiss}
                 hidden={!props.featureName}
                 gapSpace={gapSpace}
                 calloutWidth={width}
                 directionalHint={
                     props.direction === 'vertical'
                         ? DirectionalHint.bottomCenter
                         : DirectionalHint.rightCenter}>
            <div className="callout">
                <div className="header">
                    <p>{feature.name}</p>
                </div>
                {feature.description
                    ? <div className="inner">
                        <p>{feature.description}</p>
                    </div>
                    : <div className="inner empty"/>}
                <CommandBar
                    items={[
                        contextualMenuItems.featureDiagram.feature.new(feature.name, onDismiss),
                        contextualMenuItems.featureDiagram.feature.remove(feature.name, onDismiss)
                    ]}
                    farItems={[
                        contextualMenuItems.featureDiagram.feature.details(feature.name, props.onShowPanel)
                    ]}/>
            </div>
        </Callout>
    );
}