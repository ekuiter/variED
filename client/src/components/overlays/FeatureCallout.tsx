/**
 * A Fabric callout that includes information about a feature.
 */

import React from 'react';
import {Callout, DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {getSetting} from '../../store/settings';
import commands from '../commands';
import {CommandBar} from 'office-ui-fabric-react/lib/CommandBar';
import {layoutTypes, Feature} from '../../types';
import FeatureComponent, {FeatureComponentProps} from './FeatureComponent';

type Props = FeatureComponentProps & {
    onDismiss: () => void,
    isOpen: boolean,
    featureDiagramLayout: string,
    onShowOverlay: (...args: any[]) => void, // TODO
    onCollapseFeatures: (...args: any[]) => void,
    onCollapseFeaturesBelow: (...args: any[]) => void,
    onExpandFeatures: (...args: any[]) => void,
    onExpandFeaturesBelow: (...args: any[]) => void,
};

export default class extends FeatureComponent({doUpdate: true})<Props> {
    renderIfFeature(feature: Feature) {
        const {onDismiss, featureModel} = this.props,
            {gapSpace, width} = getSetting(this.props.settings, 'featureDiagram.overlay');
        return (
            <Callout target={featureModel.getElement(feature.name)!.querySelector('.rectAndText')}
                onDismiss={onDismiss}
                hidden={!this.props.isOpen}
                gapSpace={gapSpace}
                calloutWidth={width}
                directionalHint={
                    this.props.featureDiagramLayout === layoutTypes.verticalTree
                        ? DirectionalHint.bottomCenter
                        : DirectionalHint.rightCenter}>
                <div className="callout">
                    <div className="header">
                        <p>{feature.name}</p>
                    </div>
                    {feature.description
                        ? <div className="inner">
                            <p>{feature.description}</p>
                        </div>
                        : <div className="inner empty"/>}
                    <CommandBar
                        items={[
                            commands.featureDiagram.feature.newMenu(feature.name, onDismiss, true),
                            commands.featureDiagram.feature.removeMenu([feature], onDismiss, true),
                            commands.featureDiagram.feature.collapseMenu(
                                [feature], this.props.onCollapseFeatures, this.props.onExpandFeatures,
                                this.props.onCollapseFeaturesBelow, this.props.onExpandFeaturesBelow, onDismiss, true),
                        ]}
                        farItems={[
                            commands.featureDiagram.feature.details(feature.name, this.props.onShowOverlay)
                        ]}/>
                </div>
            </Callout>
        );
    }
}