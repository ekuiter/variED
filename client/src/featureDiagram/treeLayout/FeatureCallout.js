import React from 'react';
import {Callout, DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {getSetting} from '../../settings';

export default class extends React.Component {
    render() {
        return (
            <Callout target={this.props.nodeRef}
                     onDismiss={this.props.onDismiss}
                     hidden={!this.props.node}
                     gapSpace={getSetting(this.props.settings, 'featureDiagram.treeLayout.featureCallout.gapSpace')}
                     directionalHint={
                         this.props.direction === 'vertical'
                             ? DirectionalHint.bottomCenter
                             : DirectionalHint.rightCenter}>
                {this.props.node && <h3>{this.props.node.feature().name}</h3>}
            </Callout>
        );
    }
};