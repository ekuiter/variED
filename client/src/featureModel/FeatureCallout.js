import React from 'react';
import {Callout, DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import Constants from '../Constants';

export default class extends React.Component {
    render() {
        return (
            <Callout target={this.props.nodeRef}
                     onDismiss={this.props.onDismiss}
                     hidden={!this.props.node}
                     gapSpace={Constants.treeLayout.featureCallout.gapSpace}
                     directionalHint={
                         this.props.direction === 'vertical'
                             ? DirectionalHint.bottomCenter
                             : DirectionalHint.rightCenter}>
                {this.props.node && <h3>{this.props.node.feature().name}</h3>}
            </Callout>
        );
    }
};