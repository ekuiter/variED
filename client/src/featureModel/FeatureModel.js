import React from 'react';
import VerticalTreeLayout from './treeLayout/VerticalTreeLayout';
import HorizontalTreeLayout from './treeLayout/HorizontalTreeLayout';

class FeatureModel extends React.Component {
    static defaultProps = {layout: 'verticalTree'};

    static layoutMap = {
        verticalTree: VerticalTreeLayout,
        horizontalTree: HorizontalTreeLayout
    };

    layoutRef = React.createRef();

    key(props) {
        // The key uniquely identifies the layout component instance. If the key changes,
        // the instance is unmounted and a new one is mounted. This is useful for forcing
        return props.debug; // rerenders, i.e. when the debug flag changes.
    }

    render() {
        const {layout, ...props} = this.props,
            LayoutComponent = FeatureModel.layoutMap[layout];

        return (
            <LayoutComponent key={this.key(props)} ref={this.layoutRef} {...props} />
        );
    }

    export() {
        const layout = this.layoutRef.current;
        if (layout && layout.canExport())
            return layout.export().then(console.log);
        else
            return Promise.reject('can not export feature model');
    }
}

export default FeatureModel;