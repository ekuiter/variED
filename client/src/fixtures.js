
export const invalidFeatureModel1 = {};

export const invalidFeatureModel2 = {
    'struct': [],
    'constraints': [],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModel = {
    'struct': [
        {
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'his is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'name': 'JDT'},
                {'type': 'feature', 'name': 'CDT'},
                {'type': 'feature', 'name': 'AJDT'},
                {
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'name': 'CIDE'},
                        {'type': 'feature', 'name': 'FAMILIAR'},
                        {
                            'name': 'FeatureIDE',
                            'type': 'or',
                            'children': [
                                {'type': 'feature', 'name': 'AHEAD'},
                                {'type': 'feature', 'name': 'FeatureHouse'},
                                {'type': 'feature', 'name': 'FeatureCpp'},
                                {'type': 'feature', 'name': 'DeltaJ'},
                                {'type': 'feature', 'name': 'AspectJ'},
                                {'type': 'feature', 'name': 'Munge'},
                                {'type': 'feature', 'name': 'Antenna'},
                                {'type': 'feature', 'name': 'DeltaMontiArc'}
                            ]
                        },
                        {'type': 'feature', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'name': 'MoSoPoLiTe'}
                    ]
                },
                {'type': 'feature', 'name': 'DeltaJEclipsePlugin'}
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'DeltaJ'},
                        {'type': 'var', 'var': 'DeltaJEclipsePlugin'}
                    ]
                }
            ]
        }
    ],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};