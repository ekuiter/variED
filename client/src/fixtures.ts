import {SerializedFeatureModel} from './modeling/types';

/**
 * Some fixtures for use in tests.
 */

export const invalidFeatureModel1 = <SerializedFeatureModel>{};

export const invalidFeatureModel2 = <any>{
    'struct': [],
    'constraints': [],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModel = <SerializedFeatureModel>{
    'struct': [
        {
            'ID': 'Eclipse',
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'this is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'ID': 'JDT', 'name': 'JDT'},
                {'type': 'feature', 'ID': 'CDT', 'name': 'CDT'},
                {'type': 'feature', 'ID': 'AJDT', 'name': 'AJDT'},
                {
                    'ID': 'FeatureModeling',
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'ID': 'CIDE', 'name': 'CIDE'},
                        {'type': 'feature', 'ID': 'FAMILIAR', 'name': 'FAMILIAR'},
                        {
                            'ID': 'FeatureIDE',
                            'name': 'FeatureIDE',
                            'description': 'A sample description',
                            'type': 'or',
                            'children': [
                                {'type': 'feature', 'ID': 'AHEAD', 'name': 'AHEAD'},
                                {'type': 'feature', 'ID': 'FeatureHouse', 'name': 'FeatureHouse'},
                                {'type': 'feature', 'ID': 'FeatureCpp', 'name': 'FeatureCpp'},
                                {'type': 'feature', 'ID': 'DeltaJ', 'name': 'DeltaJ'},
                                {'type': 'feature', 'ID': 'AspectJ', 'name': 'AspectJ'},
                                {'type': 'feature', 'ID': 'Munge', 'name': 'Munge'},
                                {'type': 'feature', 'ID': 'Antenna', 'name': 'Antenna'},
                                {'type': 'feature', 'ID': 'DeltaMontiArc', 'name': 'DeltaMontiArc'}
                            ]
                        },
                        {'type': 'feature', 'ID': 'ExtendedFM', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'ID': 'MoSoPoLiTe', 'name': 'MoSoPoLiTe'}
                    ]
                },
                {'type': 'feature', 'ID': 'DeltaJEclipsePlugin', 'name': 'DeltaJEclipsePlugin'}
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'DeltaJ'},
                        {'type': 'var', 'var': 'DeltaJEclipsePlugin'}
                    ]
                }
            ]
        }
    ],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModelWithRemovedFeatures = <SerializedFeatureModel>{
    'struct': [
        {
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'this is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'name': 'JDT'},
                {'type': 'feature', 'name': 'CDT'},
                {'type': 'feature', 'name': 'AJDT'},
                {
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'name': 'CIDE'},
                        {'type': 'feature', 'name': 'FAMILIAR'},
                        {'type': 'feature', 'name': 'AHEAD'},
                        {'type': 'feature', 'name': 'FeatureHouse'},
                        {'type': 'feature', 'name': 'FeatureCpp'},
                        {'type': 'feature', 'name': 'AspectJ'},
                        {'type': 'feature', 'name': 'Munge'},
                        {'type': 'feature', 'name': 'Antenna'},
                        {'type': 'feature', 'name': 'DeltaMontiArc'},
                        {'type': 'feature', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'name': 'MoSoPoLiTe'}
                    ]
                }
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        }
    ],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};