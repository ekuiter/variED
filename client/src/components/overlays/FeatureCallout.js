import React from 'react';
import {Callout, DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {getSetting} from '../../store/settings';
import contextualMenuItems from '../contextualMenuItems';
import {CommandBar} from 'office-ui-fabric-react/lib/CommandBar';
import PropTypes from 'prop-types';
import {layoutTypes} from '../../types';
import {LayoutType, SettingsType} from '../../types';
import {FeatureModelType} from '../../server/FeatureModel';
import FeatureComponent from './FeatureComponent';

export default class extends FeatureComponent({doUpdate: true}) {
    static propTypes = {
        onDismiss: PropTypes.func.isRequired,
        featureModel: FeatureModelType.isRequired,
        featureName: PropTypes.string.isRequired,
        isOpen: PropTypes.bool.isRequired,
        featureDiagramLayout: LayoutType.isRequired,
        onShowOverlay: PropTypes.func.isRequired,
        onCollapseFeature: PropTypes.func.isRequired,
        onExpandFeature: PropTypes.func.isRequired,
        settings: SettingsType.isRequired
    };

    renderIfFeature(feature) {
        const {onDismiss, featureModel} = this.props,
            {gapSpace, width} = getSetting(this.props.settings, 'featureDiagram.overlay');
        return (
            <Callout target={featureModel.getElement(feature.name).querySelector('.rectAndText')}
                onDismiss={onDismiss}
                hidden={!this.props.isOpen}
                gapSpace={gapSpace}
                calloutWidth={width}
                directionalHint={
                    this.props.featureDiagramLayout === layoutTypes.verticalTree
                        ? DirectionalHint.bottomCenter
                        : DirectionalHint.rightCenter}>
                <div className="callout">
                    <div className="header">
                        <p>{feature.name}</p>
                    </div>
                    {feature.description
                        ? <div className="inner">
                            <p>{feature.description}</p>
                        </div>
                        : <div className="inner empty"/>}
                    <CommandBar
                        items={[
                            contextualMenuItems.featureDiagram.feature.new(feature.name, onDismiss),
                            contextualMenuItems.featureDiagram.feature.remove(feature.name, onDismiss),
                            contextualMenuItems.featureDiagram.feature.collapseExpand(
                                feature, this.props.onCollapseFeature, this.props.onExpandFeature, onDismiss),
                        ]}
                        farItems={[
                            contextualMenuItems.featureDiagram.feature.details(feature.name, this.props.onShowOverlay)
                        ]}/>
                </div>
            </Callout>
        );
    }
}