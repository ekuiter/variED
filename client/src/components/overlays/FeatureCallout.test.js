import React from 'react';
import {shallow} from 'enzyme';
import FeatureCallout from './FeatureCallout';
import FeatureModel from '../../server/FeatureModel';
import {validFeatureModel} from '../../fixtures';
import {layoutTypes} from '../../types';
import {defaultSettings} from '../../store/settings';
import {Callout} from 'office-ui-fabric-react/lib/Callout';

describe('FeatureCallout', () => {
    it('renders information for a feature', () => {
        const featureModel = new FeatureModel(validFeatureModel, []),
            mock = jest.fn();
        featureModel.getElement = jest.fn().mockReturnValue({
            querySelector: () => '<target element>'
        });
        const wrapper = shallow(
            <FeatureCallout
                isOpen={true}
                onDismiss={mock}
                featureDiagramLayout={layoutTypes.verticalTree}
                settings={defaultSettings}
                onShowOverlay={mock}
                onCollapseFeature={mock}
                onExpandFeature={mock}
                featureModel={featureModel}
                featureName="FeatureIDE"/>
        );
        expect(wrapper.find(Callout).find('.callout .header').contains('FeatureIDE')).toBe(true);
        expect(wrapper.find(Callout).contains('A sample description')).toBe(true);
        expect(wrapper.prop('target')).toBe('<target element>');
        expect(wrapper).toMatchSnapshot();
    });
});