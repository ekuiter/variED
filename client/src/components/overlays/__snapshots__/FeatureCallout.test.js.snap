// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureCallout renders information for a feature with a description 1`] = `
<Callout
  calloutWidth={300}
  directionalHint={5}
  gapSpace={5}
  hidden={false}
  onDismiss={[Function]}
  target="<target element>"
>
  <div
    className="callout"
  >
    <div
      className="header"
    >
      <p>
        FeatureIDE
      </p>
    </div>
    <div
      className="inner"
    >
      <p>
        A sample description
      </p>
    </div>
    <StyledCommandBarBase
      farItems={
        Array [
          Object {
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "Info",
            },
            "key": "details",
            "onClick": [Function],
            "secondaryText": "Ctrl+,",
            "text": "Details…",
          },
        ]
      }
      items={
        Array [
          Object {
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "Add",
            },
            "key": "newMenu",
            "onClick": [Function],
            "split": true,
            "subMenuProps": Object {
              "items": Array [
                Object {
                  "iconProps": Object {
                    "iconName": "Add",
                  },
                  "key": "newBelow",
                  "onClick": [Function],
                  "secondaryText": "Ctrl+N",
                  "text": "New feature below",
                },
                Object {
                  "disabled": false,
                  "iconProps": Object {
                    "iconName": "Add",
                  },
                  "key": "newAbove",
                  "onClick": [Function],
                  "text": "New feature above",
                },
              ],
            },
            "text": null,
          },
          Object {
            "disabled": false,
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "Remove",
            },
            "key": "removeMenu",
            "onClick": [Function],
            "split": true,
            "subMenuProps": Object {
              "items": Array [
                Object {
                  "iconProps": Object {
                    "iconName": "Remove",
                  },
                  "key": "remove",
                  "onClick": [Function],
                  "secondaryText": "⌫",
                  "text": "Remove feature",
                },
                Object {
                  "disabled": false,
                  "iconProps": Object {
                    "iconName": "Remove",
                  },
                  "key": "removeBelow",
                  "onClick": [Function],
                  "text": "Remove features below",
                },
              ],
            },
            "text": null,
          },
          Object {
            "disabled": false,
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "CollapseContentSingle",
            },
            "key": "collapseMenu",
            "onClick": [Function],
            "split": true,
            "subMenuProps": Object {
              "items": Array [
                Object {
                  "iconProps": Object {
                    "iconName": "CollapseContentSingle",
                  },
                  "key": "collapse",
                  "onClick": [Function],
                  "secondaryText": "Ctrl+C",
                  "text": "Collapse feature",
                },
                Object {
                  "iconProps": Object {
                    "iconName": "CollapseContent",
                  },
                  "key": "collapseBelow",
                  "onClick": [Function],
                  "text": "Collapse features below",
                },
                Object {
                  "iconProps": Object {
                    "iconName": "ExploreContent",
                  },
                  "key": "expandBelow",
                  "onClick": [Function],
                  "text": "Expand features below",
                },
              ],
            },
            "text": null,
          },
        ]
      }
    />
  </div>
</Callout>
`;

exports[`FeatureCallout renders information for a feature without a description 1`] = `
<Callout
  calloutWidth={300}
  directionalHint={5}
  gapSpace={5}
  hidden={false}
  onDismiss={[Function]}
  target="<target element>"
>
  <div
    className="callout"
  >
    <div
      className="header"
    >
      <p>
        FeatureHouse
      </p>
    </div>
    <div
      className="inner empty"
    />
    <StyledCommandBarBase
      farItems={
        Array [
          Object {
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "Info",
            },
            "key": "details",
            "onClick": [Function],
            "secondaryText": "Ctrl+,",
            "text": "Details…",
          },
        ]
      }
      items={
        Array [
          Object {
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "Add",
            },
            "key": "newMenu",
            "onClick": [Function],
            "split": true,
            "subMenuProps": Object {
              "items": Array [
                Object {
                  "iconProps": Object {
                    "iconName": "Add",
                  },
                  "key": "newBelow",
                  "onClick": [Function],
                  "secondaryText": "Ctrl+N",
                  "text": "New feature below",
                },
                Object {
                  "disabled": false,
                  "iconProps": Object {
                    "iconName": "Add",
                  },
                  "key": "newAbove",
                  "onClick": [Function],
                  "text": "New feature above",
                },
              ],
            },
            "text": null,
          },
          Object {
            "disabled": false,
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "Remove",
            },
            "key": "removeMenu",
            "onClick": [Function],
            "split": true,
            "subMenuProps": Object {
              "items": Array [
                Object {
                  "iconProps": Object {
                    "iconName": "Remove",
                  },
                  "key": "remove",
                  "onClick": [Function],
                  "secondaryText": "⌫",
                  "text": "Remove feature",
                },
                Object {
                  "disabled": false,
                  "iconProps": Object {
                    "iconName": "Remove",
                  },
                  "key": "removeBelow",
                  "onClick": [Function],
                  "text": "Remove features below",
                },
              ],
            },
            "text": null,
          },
          Object {
            "disabled": true,
            "iconOnly": true,
            "iconProps": Object {
              "iconName": "CollapseContentSingle",
            },
            "key": "collapseMenu",
            "onClick": [Function],
            "split": true,
            "subMenuProps": Object {
              "items": Array [
                Object {
                  "iconProps": Object {
                    "iconName": "CollapseContentSingle",
                  },
                  "key": "collapse",
                  "onClick": [Function],
                  "secondaryText": "Ctrl+C",
                  "text": "Collapse feature",
                },
                Object {
                  "iconProps": Object {
                    "iconName": "CollapseContent",
                  },
                  "key": "collapseBelow",
                  "onClick": [Function],
                  "text": "Collapse features below",
                },
                Object {
                  "iconProps": Object {
                    "iconName": "ExploreContent",
                  },
                  "key": "expandBelow",
                  "onClick": [Function],
                  "text": "Expand features below",
                },
              ],
            },
            "text": null,
          },
        ]
      }
    />
  </div>
</Callout>
`;
